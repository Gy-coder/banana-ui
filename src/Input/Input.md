---
title: Input
nav:
  title: Input
  path: /components
---

# Input

通过鼠标或键盘输入内容，是最基础的表单域的包装。

# 何时使用

- 需要用户输入表单域内容时。

# 代码示例

<code src="./demos/basic.tsx" />
<code src="./demos/disabled.tsx" />
<code src="./demos/readonly.tsx" />
<code src="./demos/error.tsx" />

<API src="./Input.tsx" />
