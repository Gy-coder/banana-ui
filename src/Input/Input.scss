$height: 32px;
$border-color: #999;
$border-hover-color: #666;
$border-radius: 4px;
$font-size: 12px;
$box-shadow-color: rgba(0, 0, 0, 0.5);
$red: #fa415d;
.g-input-wrapper {
  display: inline-flex;
  font-size: $font-size;
  align-items: center;
  > :not(:last-child) {
    margin-right: 0.5em;
  }
  > .g-input {
    height: $height;
    border: 1px solid $border-color;
    border-radius: $border-radius;
    padding: 0 8px;
    outline: none;
    width: 100%;
    &:hover {
      border-color: $border-hover-color;
    }
    &:focus {
      box-shadow: inset 0 1px 3px $box-shadow-color;
    }
    &[disabled],
    &[readonly] {
      border-color: #aaa;
      color: #aaa;
      cursor: not-allowed;
    }
    &.error {
      border-color: $red;
    }
  }
  &.block {
    display: flex;
  }
  .error-message {
    color: $red;
  }
}
