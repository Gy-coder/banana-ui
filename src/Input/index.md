---
title: Input
nav:
  title: Input
  path: /components
---

# Input

通过鼠标或键盘输入内容，是最基础的表单域的包装。

# 何时使用

- 需要用户输入表单域内容时。

# 代码示例

## 基础使用

<code src="./demos/basic.tsx" />

## 禁用 input

<code src="./demos/disabled.tsx" />

## input 只读

<code src="./demos/readonly.tsx" />

## 带有错误提示的 input

<code src="./demos/error.tsx" />

<API src="./Input.tsx" />
