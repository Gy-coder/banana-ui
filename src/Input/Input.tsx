import React from 'react';
import classnames from 'classnames';
import './Input.scss';

interface InputProps extends React.InputHTMLAttributes<HTMLInputElement> {
  error?: string;
}

const Input: React.FC<InputProps> = (props) => {
  const { className, readOnly, disabled, error, ...rest } = props;
  const classes = classnames('g-input', className, {
    error,
  });
  return (
    <div className="g-input-wrapper">
      <input
        className={classes}
        disabled={disabled}
        readOnly={readOnly}
        {...rest}
      />
      {error && (
        <>
          <span className="error-message">{error}</span>
        </>
      )}
    </div>
  );
};

export default Input;
