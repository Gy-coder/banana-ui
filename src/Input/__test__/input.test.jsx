import {
  act,
  cleanup,
  fireEvent,
  render,
  screen,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import React, { useState } from 'react';
import Input from '../Input';

afterEach(cleanup);

describe('测试Input组件', () => {
  it('能够正常render', () => {
    const { asFragment } = render(<Input value={'123'} onChange={() => {}} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('能够修改值', () => {
    render(
      <Input
        defaultValue="hello"
        onChange={(ev) => (document.title = ev.target.value)}
      />,
    );

    const el = screen.getByDisplayValue('hello');
    fireEvent.change(el, { target: { value: 'i love u' } });
    expect(el.value).toBe('i love u');
  });
  it('能显示错误', () => {
    render(<Input data-testid="input" error="用户名错误" />);
    expect(document.querySelector('.error-message')).toBeVisible();
  });
  it('可以disabled', () => {
    const handleChange = jest.fn();
    render(
      <Input
        defaultValue="hello"
        onChange={(e) => handleChange(e)}
        disabled={true}
      />,
    );
    const el = screen.getByDisplayValue('hello');
    expect(el).toBeDisabled();
  });
});
