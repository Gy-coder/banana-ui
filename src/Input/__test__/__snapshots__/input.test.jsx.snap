// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试Input组件 能够正常render 1`] = `
<DocumentFragment>
  <div
    class="g-input-wrapper"
  >
    <input
      class="g-input"
      value="123"
    />
  </div>
</DocumentFragment>
`;
